### PR TITLE
add fp16 for matmul

### PR DESCRIPTION
**Pull request by ForFishes**
### PR types

Bug fixes 
### PR changes

APIs
### Describe

add fp16 for matmul

### CODE REVIEW

**Review by kolinwei — DISMISSED** [review]

File: python/paddle/fluid/tests/unittests/test_matmul_v2_op.py [hunk 1]
```diff
@@ -293,7 +282,52 @@ def config(self):
...
+                     "core is not compiled with CUDA")
+    class TestMatMulOpFp16Case(parent):
+        def init_kernel_type(self):
+            self.dtype = np.float16
+
+        def test_check_output(self):
+            if core.is_compiled_with_cuda():
+                place = core.CUDAPlace(0)
+                if core.is_float16_supported(place):
+                    self.check_output_with_place(place, atol=0.2)
```
Comment (wzzju) [inline python/paddle/fluid/tests/unittests/test_matmul_v2_op.py]: Refer to https://www.johndcook.com/blog/2018/11/15/bfloat16,  I think `atol` used here is too large and `0.001` maybe the proper value.
image

Comment (ForFishes) [reply python/paddle/fluid/tests/unittests/test_matmul_v2_op.py]: 这个应该要和具体的计算相关吧。有的是相加，会很容易溢出这个精度。我看mul的设置也是非常大的。

File: python/paddle/fluid/tests/unittests/test_matmul_v2_op.py [hunk 2]
```diff
@@ -293,7 +284,54 @@ def config(self):
         self.y_shape = (100)
         self.trans_x = False
         self.trans_y = False
-        self.dtype = "float64"
+
+
+#--------------------test matmul fp16--------------------
+
+
```
Comment (wzzju) [inline python/paddle/fluid/tests/unittests/test_matmul_v2_op.py:291]: These blank lines here look ugly, do they?

**Review by wzzju — COMMENTED** [review]
LGTM . Please ask @Xreki for review.

File: python/paddle/fluid/tests/unittests/test_matmul_v2_op.py (hunk 2 above)
Comment (ForFishes) [reply python/paddle/fluid/tests/unittests/test_matmul_v2_op.py:291]: 这个是pre-commit自动空出来的。

**Review by kolinwei — APPROVED** [review]

**Review by Xreki — APPROVED** [review]
LGTM for the modification of atol and max_relative_error.

**Pull request merged**